### PR TITLE
RFC: Add YubiKey HMAC-SHA1 Challenge-Response

### PR DESCRIPTION
**Pull request by kylemanna**
## Overview

This pull request adds support for [YubiKey](https://yubico.com), a USB authentication device commonly used for 2FA.  Support is added by configuring a YubiKey slot to operate in HMAC-SHA1 challenge-response mode. The mechanism works by submitting the database master seed as a challenge to the YubiKey which replies with a HMAC-SHA1 cryptographic hash.  The resultant hash is then hashed with the other keys (password, keyfile) to generate the final key used to encrypt the database.

I have uploaded a few [screenshots to imgur](https://imgur.com/a/NbtBf).
## Details
- Breaks the ability to open the KeePass2 database by anything other then KeePassX with YubiKey support.  Perhaps KeePass2 could open it with a plugin, but I didn't look at it.
- Implemented using YubiKey HMAC-SHA1 configuration and operates completely offline.
- The master seed is used as the challenge and changes on every write to help partially combat replay attacks as it's regenerated every time the database is saved.
- YubiKey User Input Mode can be configured in a blocking or non-blocking mode.
  - Non-blocking mode is preferred such that the challenge-response can quickly be issued and returned despite being slightly less secure due to lack of user input.
  - Blocking mode works, but doesn't provide feedback to the user, so the UI appears to hang even though it's waiting.  I preferred simplicity for the initial code.
- Upon generation of the HMAC key, it can be saved and cloned to other YubiKey devices to prevent loss of database access to loss or destruction of the YubiKey hardware.  This is highly recommended.
- Only supports the first detected YubiKey Plugged in, but does enumerate both configuration slots on the device.  I only have one YubiKey and can't test multiple (yet).
- The `libyubikey` and `ykpers-1` libraries are necessary for support. These libraries are detected at compile time to enable support and necessary at run-time.
- Tested on Arch Linux and Mac OS X Mavericks with and without YubiKey support
  - Arch Linux via AUR:
    - libyubikey: 1.11-1
    - yubico-c-client-git: 2.12.r2.g539af75-1 and 2.12.r3.g262a19c-1
    - yubikey-personalization-git: 1.14.2.r5.ga7392f0-1 and 1.15.1.r10.g52890f6-1
  - Mac OS X Mavericks via Homebrew:
    - libyubikey: stable 1.11
    - ykpers: stable 1.15.1
  - Test hardware is a YubiKey Standard with firmware version 2.2.3.
- Adds challenge-response combo box to support other mechanisms in the future
- Can be disabled explicitly at compile time with `cmake -DCMAKE_DISABLE_FIND_PACKAGE_YubiKey=TRUE`
## Quick Start Guide
1. Install necessary libraries to appease KeePassX's CMake configuration step
   
   ```
   -- The following OPTIONAL packages have been found:
   * YubiKey
   ```
2. Install [yubikey-personalization-gui](https://github.com/Yubico/yubikey-personalization-gui) and configure slot 2 for HMAC-SHA1 challenge-response.  Be sure to write the configuration.  Optionally it can be programmed via the command line.
3. Start up KeePassX with the YubiKey inserted and the challenge-response form of the DatabaseOpenWidget ChangeMasterKeyWidget should enable themselves asynchronously listing the inserted YubiKey.  Review [screenshots on imgur](https://imgur.com/a/NbtBf) for the configuration tool.
## Upstream / Reference
- [yubico open source software](https://www.yubico.com/develop/open-source-software/)
- [libyubikey](https://github.com/Yubico/yubico-c)
- [yubico-c-client](https://github.com/Yubico/yubico-c-client)
- [yubikey-personalization-gui](https://github.com/Yubico/yubikey-personalization-gui)

### CODE REVIEW

**Comment by piepkrak** [comment]
Works like a charm. Thank you @kylemanna.

Got it running on OS X 10.10 Yosemite. Installed Qt 4.8.7 from source. Had to configure Qt with -no-phonon as phonon caused make errors.

I also had to change find_path statements to find the Yubikey libraries as they were installed with macports. They are probably not in the PATH? I'm not familiar with cmake.
```
-find_path(YUBIKEY_CORE_INCLUDE_DIR yubikey.h)
-find_path(YUBIKEY_PERS_INCLUDE_DIR ykcore.h PATH_SUFFIXES ykpers-1)
+find_path(YUBIKEY_CORE_INCLUDE_DIR yubikey.h PATHS /opt/local/include)
+find_path(YUBIKEY_PERS_INCLUDE_DIR ykcore.h PATHS /opt/local/include PATH_SUFFIXES ykpers-1)
 set(YUBIKEY_INCLUDE_DIRS ${YUBIKEY_CORE_INCLUDE_DIR} ${YUBIKEY_PERS_INCLUDE_DIR})

-find_library(YUBIKEY_CORE_LIBRARY yubikey)
-find_library(YUBIKEY_PERS_LIBRARY ykpers-1)
+find_library(YUBIKEY_CORE_LIBRARY yubikey PATHS /opt/local/lib)
+find_library(YUBIKEY_PERS_LIBRARY ykpers-1 PATHS /opt/local/lib)
 set(YUBIKEY_LIBRARIES ${YUBIKEY_CORE_LIBRARY} ${YUBIKEY_PERS_LIBRARY})
```

**Comment by kylemanna** [comment]
>Got it running on OS X 10.10 Yosemite. Installed Qt 4.8.7 from source. Had to
>configure Qt with -no-phonon as phonon caused make errors.
>
>I also had to change find_path statements to find the Yubikey libraries as they
>were installed with macports. They are probably not in the PATH? I'm not
>familiar with cmake.
>
>-find_path(YUBIKEY_CORE_INCLUDE_DIR yubikey.h)
>-find_path(YUBIKEY_PERS_INCLUDE_DIR ykcore.h PATH_SUFFIXES ykpers-1)
>+find_path(YUBIKEY_CORE_INCLUDE_DIR yubikey.h PATHS /opt/local/include)
>+find_path(YUBIKEY_PERS_INCLUDE_DIR ykcore.h PATHS /opt/local/include PATH_SUFFIXES ykpers-1)
> set(YUBIKEY_INCLUDE_DIRS ${YUBIKEY_CORE_INCLUDE_DIR} ${YUBIKEY_PERS_INCLUDE_DIR})
>
>-find_library(YUBIKEY_CORE_LIBRARY yubikey)
>-find_library(YUBIKEY_PERS_LIBRARY ykpers-1)
>+find_library(YUBIKEY_CORE_LIBRARY yubikey PATHS /opt/local/lib)
>+find_library(YUBIKEY_PERS_LIBRARY ykpers-1 PATHS /opt/local/lib)
> set(YUBIKEY_LIBRARIES ${YUBIKEY_CORE_LIBRARY} ${YUBIKEY_PERS_LIBRARY})

Good to hear, I wonder if you can work around this with something like
the CMAKE_PREFIX_PATH environemental variable?  I can't check right now,
but if you ever rebuild, might check it out.

I didn't have to do any of that because my Mac tests used homebrew for
Qt and yubikey libraries.  Just worked.  I was shocked since I did the
devel in Linux.

**Comment by kylemanna** [comment]
@debfx Any comments? Any interest in merging?

**Comment by c8c-erzatz-info** [comment]
I'm waiting this feature for a long time! Just cloned, installed and it works.

For some reason, when I save it doen't add .kdb extension.

Thanks!

**Comment by kylemanna** [comment]
@c8c-erzatz-info Do you have the file extension problem with the master
code branch?  I need to merge the recent upstream changes to my repo and
branch if you were using that.

As for other news, I think I'm going to fork this in the near future and
maintain the YubiKey code.  This is the most commented open pull request
on this repository with no feedback (good or bad) regarding merging from
the maintainer / owner.

Other thoughts from those interested?

**Comment by c8c-erzatz-info** [comment]
@kylemanna Sorry for this thread pollution. The main branch (835c411d1256ee3527c38916e1fd7bd323eeb799) and it has the same behaviour. So your patch doesn't cause this issue.

It's too bad one have to fork this project to have access to yubikey compatibility. I'll split contributions between two projects. But I understand that it might be the only option.

**Comment by pgalves** [comment]
Works very well, nice work!
@kylemanna I just have a minor glitch. I don't know if it is expected to work like it is, but I'm unable to remove the Yubikey protection from the database. When changing passwords and disabling the Yubikey challenge combo and saving still requires the Yubikey when opening the database. I made a quick fix, you can find it here: https://github.com/pgalves/keepassx/commit/7a7e919f8ff35f50a95fec32df1b235e3f037eb1.

**Comment by FrostbittenKing** [comment]
:+1: Also would like to have this feature implemented. Just tested it on arch linux and made a PKGBUILD for it.

**Comment by kylemanna** [comment]
> My only hesitation with adopting this as-is is related to being able to use my database an Android as well as from Linux/Mac. I understand some Yubikeys have NFC features. I wonder .... is this implementation going to play nice with future attempts to offer hardware key support in other programs? Have you been approached by anyone with other implementation schemes?

@alerque it appears that a challenge response can be issued over NFC as an [Android app](https://play.google.com/store/apps/details?id=com.yubichallenge) exists claiming to do that.  There's no reason Linux + Mac can't use the USB challenge response, in fact I've tested both.

**Comment by kylemanna** [comment]
@pgalves Thanks for the catch and patch!  I've merged it.  Have a pat on the back @changetip

**Comment by kylemanna** [comment]
@FrostbittenKing you inspired me to create my first AUR:

https://aur.archlinux.org/packages/keepassx2-yubikey-git

I'll keep it up to date as I make changes to YubiKey patches and merge upstream/master.

**Comment by FrostbittenKing** [comment]
@kylemanna, awesome. Did something stupid. Built the PKGBUILD in /tmp. Well, shiiiite, after reboot .. of course it was gone. Little suggestion. I did it a bit differently. I added the original keepassx repo and in the prepare() function, I added your repo as additonal remote source, and afterwards merged your branch with it.
Something like this:
```
_check_repo_added() {
     git show-ref --verify --quiet refs/remotes/keepassx-yubikey/yubikey
     return $?
}

prepare() {
    if ! _check_repo_added
    then
        git remote add -t yubikey keepassx-yubikey git@github.com:kylemanna/keepassx.git
        git fetch keepassx-yubikey
        git merge --no-ff --no-edit keepassx-yubikey/yubikey
     fi
}
```

**Comment by kylemanna** [comment]
@FrostbittenKing Thanks for the trick.  I integrated your suggestion.  It's annoying that `pkgver()` is called before `prepare()` which results in an incorrect `pkgver`.  Sigh. Nevertheless, I'll accept it for now.  If you have any more comments w.r.t. the `PKGBUILD`, post a comment on AUR there so that we don't create too much Arch noise here.

**Comment by pgalves** [comment]
I've implemented a simple CLI tool (utils/kdbx-noyubikey-exporter) that can be used to unlock a Yubikey protected database without the physical Yubikey.
You need to have a backup of the (40 hex char) secret  key used to configure the Yubikey. The tool just opens the database using the given password/keyfile + computed HMAC-SHA1 using the 40 char hex secret key and than writes a copy of this database without the Yubikey protection. It keeps the password and/or keyfile of the original database. You can find the the tool in this commit: 69f71cbf3c6fa12c8183f7b69d64f470dc46d51a

The tool can be useful in an emergency, if someone only has one Yubikey or needs the database in device that does not supports the Yubikey. Be aware that having a backup of the Yubikey secret key decreases the security. The advantage of the hardware token is that secret key never leaves the device but if you have an insecure backup of it...

Feedback is welcome.
@kylemanna sorry for the noise here, you already said that a tool like this was out of scope of this PR.

**Comment by kylemanna** [comment]
@pgalves That's awesome!  As for noise in this PR, originally I was hoping this would get merged and wanted to stay focused.  Instead YK support looks more like a fork.  That said, you should definitely open a PR on my repo so we can merge.  It'd be nice to add some unit tests and get some feedback from the others subscribed to this PR.

**Comment by FrostbittenKing** [comment]
@kylemanna, did you get any response from debfx? At least I'd like to know, why debfx basically ignores your pull request. Another thing I need to know is, if there is some way to use the yubikey-patched keepassxdb on a android device? I love your fork, but as long as I cannot use this on android (or other people on ios ..if you will), it still pretty much useless to me :(. On the other hand, this might be the perfect starting point for me to work on a community-driven OSS project.

**Comment by kylemanna** [comment]
@FrostbittenKing -- @debfx has never responded ever.  Not sure why, this is the most commented pull request ever, so I'm sure he's seen it.  Could be worse, could charge in here and close it. :)  I interpret no response after almost a year as "no".

As for Android devices, with a YubiKey Neo people can do challenge-response over NFC. See [my post above](https://github.com/keepassx/keepassx/pull/52#issuecomment-71848026).  Just needs someone to hack at it.

**Comment by jkulesza** [comment]
@kylemanna, I submitted a trivial pull request to clean up the README and have seen no reply/response/action from @debfx either.  I'm concerned that development is stagnating even at the alpha stage and that new users will be turned off by that...

**Comment by pgalves** [comment]
@kylemanna thanks! I just opened a PR in your repo https://github.com/kylemanna/keepassx/pull/6 for the emergency tool to unlock the database. I just added some tests since my last commit.

If the others here can give it a try, I can improve it.

**Comment by FrostbittenKing** [comment]
@kylemanna. I read that post, or rather I misread it. I thought it was working and tried it, only to fail, and thought .. bleh no it is not working. Thanks for clearing that up.

**Comment by unaizalakain** [comment]
Is there any will of maintaining this branch updated with master and accepting pull requests to it?

**Comment by kylemanna** [comment]
@unaizalakain All development is continuing on my repo @ kylemanna/keepassx .  There has been no response for or against merging this feature.

**Comment by nickjones** [comment]
Are there plans to merge this into master still or is there another pull request that cleanly merges where progress is being made?  I'd really like to see this make it in.

**Comment by kylemanna** [comment]
> Are there plans to merge this into master still or is there another pull request that cleanly merges where progress is being made? I'd really like to see this make it in.

I've forked it on my repository and others have been using it.  It builds cleanly on old code and I haven't rebased it recently because I haven't had time and no major release milestones have been reached.

https://github.com/kylemanna/keepassx

**Comment by nickjones** [comment]
@kylemanna, thanks for the info.  I made a rebased copy and fixed the merge conflicts that also seems to work.  Thanks for putting this together.

**Comment by salibhai1** [comment]
Although I dont use KeepassX , I think this is an awesome idea and helps further the cause of security for everyone!  Keep up the good work

@ChangeTip, send a pie to @kylemanna

**Comment by kylemanna** [comment]
@salibhai1 thanks!  Too bad @ChangeTip bot doesn't agree, about par for the course from my experience.

**Comment by salibhai1** [comment]
strange... okay lets try again. @changetip send $3.14 to @kylemanna

**Comment by BashCo** [comment]
Sorry the bot isn't working! Let me try. @changetip send $1 to @salibhai1

**Comment by BashCo** [comment]
Trying again after a fix was pushed. Sorry for the mishap! @changetip send $1 to @salibhai1.

**Comment by MaxFangX** [comment]
@changetip test

**Comment by MaxFangX** [comment]
@changetip see me please a coffee

**Comment by salibhai1** [comment]
@ChangeTip, send a pie to @kylemanna

**Comment by MaxFangX** [comment]
@changetip thanks

**Comment by changetip** [comment]
Hi @kylemanna, I've delivered a Bitcoin tip worth a coffee (4,979 bits/$1.50) from @MaxFangX to your ChangeTip wallet.

**[Learn about ChangeTip](https://www.changetip.com/tip-online/github)**

**Comment by changetip** [comment]
Hi @kylemanna, I've delivered a Bitcoin tip worth 1 thanks (2,490 bits/$0.75) from @MaxFangX to your ChangeTip wallet.

**[Learn about ChangeTip](https://www.changetip.com/tip-online/github)**

**Comment by changetip** [comment]
Hi @kylemanna, I've delivered a Bitcoin tip worth a pie (10,423 bits/$3.14) from @salibhai1 to your ChangeTip wallet.

**[Learn about ChangeTip](https://www.changetip.com/tip-online/github)**

**Comment by MaxFangX** [comment]
@salibhai1 Really sorry for all the delays in tipping through ChangeTip. The Github API only returns 30 comments per API call, and our bot didn't detect your comment because this was a particularly long comment thread. Have a beer on me for being the unfortunate recipient of this bug. Cheers! @changetip

**Comment by changetip** [comment]
Hi @salibhai1, I've delivered a Bitcoin tip worth a beer (11,618 bits/$3.50) from @MaxFangX to your ChangeTip wallet.

**[Learn about ChangeTip](https://www.changetip.com/tip-online/github)**

**Comment by tristan-k** [comment]
Does someone have a working build for Mac OS X?

**Comment by kylemanna** [comment]
@tristan-k I had compiled it on my old Mac just fine, but I no longer have that machine so can't confirm.  The branch on my repo if it doesn't just work is certainly very close.

**Comment by tristan-k** [comment]
Does your version support the KeePass 2.x database format, because I'm getting a strange error?

ishot-2

**Comment by kylemanna** [comment]
It does support keepass2, also i have no clue what your error says.

**Comment by tristan-k** [comment]
@kylemanna never mind. the error is related to the KeeChallenge incompatibility. Your fork is working great otherwise.

**Comment by kylemanna** [comment]
@tristan-k Glad to hear you got it working! :smile:  :key: :lock:

**Comment by KFDCompiled** [comment]
@tristan-k i've had trouble with keechallenge incompatibility as well, but the XML files accompanying the kdbx files are structured identically:
```
<?xml version="1.0" encoding="UTF-8"?>
<data>

    <aes>
        <encrypted> . . . </encrypted>
        <iv> . . . </iv>
    </aes>

    <challenge> . . . </challenge>
    <verification> . . . </verification>
    <lt64> . . . </lt64>

</data>
```
@kylemanna Thanks for the awesome work! Is this anything you can shed some light on?

**Comment by kylemanna** [comment]
> @kylemanna  Is this anything you can shed some light on?

@KFDCompiled The files are not compatible and never will be.  KeeChallenge needs a separate file to store the challenge and response likely as a result of the plugin architecture.  With my patch it's all integrated in one file.  Less to manage, lose, corrupt, etc.

**Comment by KFDCompiled** [comment]
@kylemanna If I understand correctly, KeeChallenge stores the challenge and response in its accompanying xml file but your patch does not. I'm having trouble understanding how your patch uses the xml file and why to use an external file at all—couldn't the information be stored in a user-editable (at their peril) database entry?

**Comment by kylemanna** [comment]
My patch doesn't use the xml file.  It doens't work with KeeChallenge.  They are different implementations.

The data can't be stored in the database because 1) the entries in the database are encrypted so we couldn't read the challenge ahead of time and 2) adding and unencrypted field would create a new keepass file format.

**Comment by KFDCompiled** [comment]
>My patch doesn't use the xml file.

@kylemanna I checked and you are, of course, correct. I do not know where that file came from and why I though it came from your patch. It all makes sense now :suspect:

**Comment by skrzepto** [comment]
Any plans to merge with master?

**Comment by abuisman** [comment]
What is the status now? Would be cool if keepassx supported yubikey

**Comment by mongrelion** [comment]
@kylemanna first than anything, thanks for the hard work!
I just read the whole thread and it is not clear to me (and apparently to other people) if whether or not this feature is going to be merged into master. Who's in charge of that?
The fact that your fork is working is good news but unfortunately it also means outdated code (as you said you haven't had time to rebase with mainstream) which means no bug nor security fixes since you forked :(

**Comment by kylemanna** [comment]
@mongrelion It will probably never get merged.  The maintainer (@debfx) has never commented on this thread despite being asked.  It's the largest pull request by comments on the project, so I'm sure he is aware. Adding hardware support to an otherwise hardware independent piece of software is a big ask.  I assume the maintainer isn't interested in hauling around this code.  His call.

As for rebasing, you're right.  I just don't have the time right now, and I've yet to see any major features get added that necessitate it.  I'm waiting for the project to enter beta or some milestone.  Rebasing will be time consuming from a testing standpoint, and I don't want to create a build that corrupts somebody's password database because I rushed.

If people are interested in discussing the fork, or if there is a critical mass of people looking for the rebase (and willing to test!), please take the conversation to the fork and appropriate issue there.

https://github.com/kylemanna/keepassx

**Comment by kylemanna** [comment]
@mongrelion Just rebased the fork.  Give it a test now.  If you discover any issues, please create issues on my [kylemanna/keepassx](https://github.com/kylemanna/keepassx) repo.

The old fork branch head moved to yubikey-old in the event anyone has forks of my fork.

**Comment by kerberizer** [comment]
If @debfx wants to keep the core KeePassX code as clean and lean as possible, this is actually quite understandable: KeePassX is a rather critical application and, being a non-commercial open source project, developer support may at times be extremely limited. One solution may be a plugin architecture and in another discussion debfx did mention that he planned to work on this once 2.0 is released. So, considering that the YubiKey functionality would indeed be a very welcome and helpful addition (not to mention that it would be a pity to ditch all the work already done on it or to dissipate the already limited development resources on different forks), I wonder if it would be possible to (re)implement this functionality as a plugin? Probably you guys should liaise on this matter?

**Comment by Manko10** [comment]
Very cool. I guess I'll build a PKGBUILD for it right away.

BTW something seems to be wrong with the CI checks, but it looks like it's just some kind of name resolution issues with github.com, nothing in the code itself.

**Comment by alerque** [comment]
@Manko10 There already is a PKGBUILD — and since it pull from git HEAD it should be ready to go. If you reinstall it it should be using the latest sources as mentioned above.

**Comment by Manko10** [comment]
Yeah, right, I found it. Thanks.

**Comment by Manko10** [comment]
I built it and it's working. However, I found three issues:

1. When I select a database to open and then plug in the YubiKey, the Challenge Response option stays grayed out. I have to plug in the YubiKey first.
2.  Presence of the YubiKey is checked every time I try to open the database which takes a while. Unfortunately, KeePassX doesn't remember whether I used a YubiKey last time, so every time I want to open a database I have to wait for the YubiKey to be recognized and then activate "Challenge Response" manually before I can open the database. This is pretty annoying
3. Sometimes the dropdown shows my YubiKey several times (sometimes just twice, sometimes as often as 6 times or more)

**Comment by kylemanna** [comment]
@Manko10 

> When I select a database to open and then plug in the YubiKey, the Challenge Response option stays grayed out. I have to plug in the YubiKey first.

I omitted the extra work of autodetection that would allow you to hot plug the Yubikey.  It'd be more code that I'd have to port for every release.  It's on the list of things to do should the initial patch get merged.  This was discussed [in the past as well](https://github.com/keepassx/keepassx/pull/52#issuecomment-46699399).

> Presence of the YubiKey is checked every time I try to open the database which takes a while. Unfortunately, KeePassX doesn't remember whether I used a YubiKey last time, so every time I want to open a database I have to wait for the YubiKey to be recognized and then activate "Challenge Response" manually before I can open the database. This is pretty annoying

A YubiKey service that listens hot-plug events would resolve this.  The current approach is *KISS* focusing on the crypto and database robustness.  Polish such as hot-plug support similar to the YubiKey personalization app could be added later.

> Sometimes the dropdown shows my YubiKey several times (sometimes just twice, sometimes as often as 6 times or more)

This is a bug that I've never seen it.  If you could open an issue on my [fork](https://github.com/kylemanna/keepassx) with info about the Yubico libraries in use and and any debugging info.  running `cmake -DCMAKE_BUILD_TYPE=Debug` would cause some output regarding challenge-response to be printed to the console, perhaps this might help highlight the bug.

**Comment by Manko10** [comment]
> A YubiKey service that listens hot-plug events would resolve this. The current approach is KISS focusing on the crypto and database robustness. Polish such as hot-plug support similar to the YubiKey personalization app could be added later.

Well, my easy lazy implementation would have been to simply add one or two lines of extra code to the part that is enabling the dropdown. If the option was enabled at last login, enable it again. Otherwise leave it unchecked. This still doesn't get rid of the delay, but that is not so critical since I need time for entering the password anyway. But at least I don't need to manually click the checkbox anymore.

> This is a bug that I've never seen it. If you could open an issue on my fork with info about the Yubico libraries in use and and any debugging info. running cmake -DCMAKE_BUILD_TYPE=Debug would cause some output regarding challenge-response to be printed to the console, perhaps this might help highlight the bug.

Sure, I'll do that.

**Comment by juju4** [comment]
+1
seems all issues where solve? can it be merged?

Thanks!

**Comment by skorokithakis** [comment]
+1 on this, any updates?

**Comment by Ardakilic** [comment]
+1, I'd love to have this feature implemented.

**Comment by sanyo-ok** [comment]
Hello,

I am in the process of evaluating different methods of securing my KeepAss password store.
OtpKeyProv is looking interesting, but I need to fully understand all dependencies
beetween keys and different conditions before using it with a real KP database.

My main concerns:
1) To avoid being locked out from my own KP store
2) To be sure it is secure enough,
that at least other people cannot extract secret key from a stolen KP database,
otherwise it is not useful to bother with all this typing these 3 pins everyday.

If I understand correctly verification of OTP is done on both sides (client and server)
by calculating following function:
HMAC(SecretKey,Counter)
So the SecretKey should be present on both client and server to allow client
to generate one time password and to allow server to verify if it is
a correct one time password received from a client.

Please help me to understand where secret key is stored?

After some reading I have found OtpKeyProv is somehow related to key file feature
of the KeepAss application. I guess may be a product of OTP HMAC calculation is used
to generate an analogue of temporary key file to unlock KP database?
But I do not understand where secret key is taken for calculation of HMAC by the plugin, especially taking into account secret shall be entered manually during recovery.

If the secret key is not present withing the KP database, then how HMAC(Key,Count) is calculated by the plugin?

But if the secret key is present in the database or in another place on the computer then it is not secure at all and the only another secure method is challenge-response with assymetric keys?

**Comment by Canalytic** [comment]
Im using KeePassX 0.4.3 for Ubuntu Linux.  I cannot find the challenge response settting for the master key, as per your first uploaded image.

**Comment by polyzen** [comment]
@Canalytic, this pull request wasn't merged. Afaik you have to use his fork for this feature.

**Comment by kerberizer** [comment]
@Canalytic, also, this repo is for KeePassX 2.0. KeePassX 0.4.3 is the (rather) older previous version.

**Comment by Scoubi** [comment]
Hi, 

I manage to compile this fork on Mac OS X (El Capitan)

I'm unable to open my keepass db using my Yubiley. However, if I create a new DB with KeePassX with the challenge response it works. 

I know my kdbx is good because I use it with Keepass2droid and Keepass on Windows. 

Any idea on why it's not working and how I can fix it?
I used the latest libyubikey (1.13) and the latest ykpers (1.17.3) which are not the one refered to in this thread.
I also had to use qt5 and not qt4-mac

One thing I noticed, when I created my kdbx with keepass it also created a .xml file. When I created it with KeePassX it didn't.

I really want to be able to use my kdbx on my mac. 

Thanks.

**Comment by denibertovic** [comment]
Is there still no comment from the maintainers on this? What needs to happen to get this merged and released?

**Comment by Nairwolf** [comment]
@denibertovic : Look when was the last commit of @debfx (the maintainer, I suppose), 25 May. I think he's really busy. 

Also, a lot of others PR hasn't been merged yet...

I don't know if he accepts PR by the way.

**Comment by TheZ3ro** [comment]
News?
This PR Need a Rebase

**Comment by sondr3** [comment]
You could try to create this pull request over at https://github.com/keepassxreboot/keepassx

**Comment by ytisf** [comment]
Willing to support this feature with cash as well. Please update.

File: src/gui/IconModels.h [hunk 1]
```diff
@@ -41,14 +41,14 @@ class CustomIconModel : public QAbstractListModel
 public:
     explicit CustomIconModel(QObject* parent = nullptr);
 
-    int rowCount(const QModelIndex& parent = QModelIndex()) const override;
-    QVariant data(const QModelIndex& index, int role = Qt::DisplayRole) const override;
-    void setIcons(const QHash<Uuid, QImage>& icons, const QList<Uuid>& iconsOrder);
+    virtual int rowCount(const QModelIndex& parent = QModelIndex()) const Q_DECL_OVERRIDE;
```
Comment (TheZ3ro) [inline src/gui/IconModels.h]: Why the virtual keyword and the `Q_DECL_OVERRIDE` instead of C `override` ?

Comment (kylemanna) [inline src/gui/IconModels.h]: Great question, I don't recall.  Likely two minor bugs.  I think you're the first person to review the code. Ever. :)

**Comment by sami10007** [comment]
I requested a review about your code from YubiKey support team. They say 

> We do not support [KeePassX]. We do know that they have already added support for CR several years ago. As to supporting the developer in getting this working, if they reach out to us we are more than happy to work with them on this.

Please, contact Aaron of the support team for troubleshooting the case further. It would be nice to get some reviews about the condition of the ticket.

**Comment by TheZ3ro** [comment]
@sami10007 this feature has been merged in KeePassXC (the community fork of KeePassX) here: https://github.com/keepassxreboot/keepassxc/pull/127, we talked with @a-dma in the PR and we will be happy to discuss further

**Comment by kylemanna** [comment]
Thanks @TheZ3ro and others on the KeePassXC team for giving this a new home.  Closing issue.

**Pull request closed without merging**